geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java

```diff
@@ -1215,25 +1216,26 @@ public List peek(int batchSize, int timeToWait) throws InterruptedException, Cac
     final boolean isDebugEnabled = logger.isDebugEnabled();
 
     PartitionedRegion prQ = getRandomShadowPR();
-    List<GatewaySenderEventImpl> batch = new ArrayList<>();
     if (prQ == null || prQ.getLocalMaxMemory() == 0) {
       try {
         Thread.sleep(50);
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
       }
       blockProcessorThreadIfRequired();
-      return batch;
+      return Collections.EMPTY_LIST;
     }
+    List<GatewaySenderEventImpl> batch =
+        new ArrayList<>(batchSize == BATCH_BASED_ON_TIME_ONLY ? DEFAULT_BATCH_SIZE : batchSize);
 
     long start = System.currentTimeMillis();
     long end = start + timeToWait;
 
     // Add peeked events
-    addPeekedEvents(batch, batchSize);
+    addPeekedEvents(batch, batchSize == BATCH_BASED_ON_TIME_ONLY ? DEFAULT_BATCH_SIZE : batchSize);
 
     int bId = -1;
-    while (batch.size() < batchSize) {
+    while (batchSize == BATCH_BASED_ON_TIME_ONLY || batch.size() < batchSize) {
```

Comment:
Is there a need to control; how big a batch size can grow...If user has set a long time, the batch size could be huge.

Reply:
This is for a specific case where an application just wanted it time triggered.  The work around is to specify a large batch-size currently but that also has large overhead.  In this case, I believe the application developers can determine what an ideal time is.